File: 백준/Gold/20055. 컨베이어 벨트 위의 로봇/컨베이어 벨트 위의 로봇.py

# 0. 입력값
N,K = map(int,input().split())
belt = list(map(int,input().split()))

robot = [False] * N
size = N*2
stage = 0

while True :
    stage += 1

    # 1-1. 벨트 한바퀴 회전
    temp = belt[size-1]
    for i in range(size-1,0,-1) :
        belt[i] = belt[i-1]
    belt[0] = temp

    # 1-2. 로봇 함께 회전
    for i in range(N-1, 0, -1) :
        robot[i] = robot[i-1]
    robot[0] = False        
    robot[N-1] = False      # 끝에 도착하면 로봇 빼기

    # 2. 가장 먼저 벨트에 올라간 로봇부터, 벨트가 회전하는 방향으로 한 칸 이동(이동칸에 로봇이 없고, 내구도가 1 이상일 경우에만 이동)
    for i in range(N-1, 0, -1) :
        if robot[i-1] and robot[i] == False and belt[i] > 0 :
            robot[i] = robot[i-1]
            robot[i-1] = False
            belt[i] -= 1
    
    robot[N-1] = False      # 끝에 도착하면 로봇 빼기

    # 3. 로봇 올리기 (내구도 1이상일 경우)
    if belt[0] > 0 :
        robot[0] = True
        belt[0] -= 1

    # 4. 내구도가 0인 칸의 개수가 K개 이상이면 종료, 아니면 1번으로 이동 
    if belt.count(0) >= K :
        break

print(stage)

File: 백준/Gold/20055. 컨베이어 벨트 위의 로봇/컨베이어 벨트 위의 로봇.java

import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        // 0. 값 설정
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(st.nextToken());

        st = new StringTokenizer(br.readLine());
        int[] belt = new int[N*2];
        for (int i = 0; i < N*2; i++) {
            belt[i] = Integer.parseInt(st.nextToken());
        }
        boolean[] robot = new boolean[N];
        int stage = 0;

        while (true) {
            stage++;

            // 1-1. 벨트 회전
            int tmp = belt[(N*2)-1];
            for (int i = (N*2)-1; i > 0 ; i--) {
                belt[i] = belt[i-1];
            }
            belt[0] = tmp;

            // 1-2. 로봇 회전
            for (int i = N-1; i > 0; i--) {
                robot[i] = robot[i-1];
            }
            robot[0] = false;

            robot[N-1] = false;     // 끝 도달 로봇 제거

            // 2. 로봇만 앞으로 이동
            for (int i = N-1; i >0; i--) {
                if (robot[i-1] && !(robot[i]) && belt[i]>0) {
                    robot[i] = robot[i-1];
                    robot[i-1] = false;
                    belt[i]--;
                }
            }
            robot[N-1] = false;     // 끝 도달 로봇 제거

            // 3. 로봇 올리기
            if (!robot[0] && belt[0]>0) {
                robot[0] = true;
                belt[0]--;
            }

            // 4. 내구도 확인
            int cnt = 0;
            for (int i = 0; i < (N*2); i++) {
                if (belt[i] == 0) {
                    cnt++;

                    if (cnt >= K) {
                        break;
                    }
                }
            }
            if (cnt >= K) {
                break;
            }

        }

        System.out.println(stage);
    }

}